packages/fluentui/react-northstar/src/components/Dropdown/Dropdown.tsx

```diff
@@ -826,31 +817,154 @@ class Dropdown extends AutoControlledComponent<WithAsProp<DropdownProps>, Dropdo
+    const { search, multiple, highlightFirstItemOnOpen, items, getA11ySelectionMessage } = this.props;
+    const { value, open } = this.state;
+    const {
+      type,
+      inputValue: newSearchQuery,
+      selectedItem: newValue,
+      isOpen: newOpen,
+      highlightedIndex: newHighlightedIndex,
+    } = changes;
+    const newState = {} as DropdownState;
+    const handlers = [];
+
+    switch (type) {
+      case Downshift.stateChangeTypes.changeInput: {
+        const shouldValueChange = newSearchQuery === '' && !multiple && !!value.length;
+        newState.searchQuery = newSearchQuery;
+        newState.highlightedIndex = highlightFirstItemOnOpen ? 0 : null;
+
+        if (shouldValueChange) {
+          newState.value = [];
         }
-        if (!this.props.search) {
-          this.listRef.current.focus();
+
+        if (open) {
+          // we clear value when in single selection user cleared the query.
+          const shouldMenuClose = newSearchQuery === '' || newValue !== undefined;
+
+          if (shouldMenuClose) {
+            newState.open = false;
+          }
+        } else {
+          newState.open = true;
         }
-      } else {
-        newState.highlightedIndex = null;
+
+        break;
       }
+      case Downshift.stateChangeTypes.keyDownEnter:
+      case Downshift.stateChangeTypes.clickItem:
+        const shouldAddHighlightedIndex = !multiple && !!items;
+
+        newState.searchQuery = this.getSelectedItemAsString(newValue);
+        newState.value = multiple ? [...value, newValue] : [newValue];
+        newState.open = false;
+
+        if (shouldAddHighlightedIndex) {
+          newState.highlightedIndex = items.indexOf(newValue);
+        }
+
+        if (getA11ySelectionMessage && getA11ySelectionMessage.onAdd) {
+          this.setA11ySelectionMessage(getA11ySelectionMessage.onAdd(newValue));
+        }
+
+        if (multiple) {
+          setTimeout(() => (this.selectedItemsRef.current.scrollTop = this.selectedItemsRef.current.scrollHeight), 0);
+        }
+
+        this.tryFocusTriggerButton();
+
+        break;
+      case Downshift.stateChangeTypes.keyDownEscape:
+        if (search) {
+          newState.searchQuery = '';
+
+          if (!multiple) {
+            newState.value = [];
+          }
+        }
+        newState.open = false;
+        newState.highlightedIndex = highlightFirstItemOnOpen ? 0 : null;
+        break;
+      case Downshift.stateChangeTypes.keyDownArrowDown:
+      case Downshift.stateChangeTypes.keyDownArrowUp:
+        if (newOpen !== undefined) {
+          newState.open = newOpen;
+          newState.highlightedIndex = newHighlightedIndex;
+
+          if (newOpen) {
+            const highlightedIndexOnArrowKeyOpen = this.getHighlightedIndexOnArrowKeyOpen(changes);
+
+            if (_.isNumber(highlightedIndexOnArrowKeyOpen)) {
+              newState.highlightedIndex = highlightedIndexOnArrowKeyOpen;
```

Comment:
With [these changes](https://github.com/microsoft/fluentui/pull/12361/files#r397094666) we can simplify:

```suggestion
            newState.highlightedIndex = this.getHighlightedIndexOnArrowKeyOpen(changes);
```

Reply:
it will set state to undefined for highlightedIndex, not what we want, but most probably this code can be improved. For this PR I just tried to bring all the handling for state change of Downshift into one place, so it can be read better. Maybe we will improve the code in the future, for now I think it's good enough.